### ISSUE TITLE
XpsDocument.GetFixedDocumentSequence throws NotSupportedException

### ISSUE THREAD

**Issue by elyoh** [dotnet/wpf#1063]
•    NET Core Version: 3.0.0-preview6-27804-01 (also in preview 5).
•    Windows version: Win10
•    Does the bug reproduce also in WPF for .NET Framework 4.8?: No

An attempt is made to read an XPS document from disk:

```
var doc = new XpsDocument("path to file", FileAccess.Read);
var seq = doc.GetFixedDocumentSequence();
```

The following exception is thrown:
```
System.Windows.Markup.XamlParseException: ''Initialization of 'System.Windows.Documents.FixedDocument' threw an exception.' Line number '1' and line position '479'.'
```
Inner exception:
```
NotSupportedException: Specified method is not supported.
```

Stack trace:
```
at System.Windows.Markup.XamlReader.RewrapException(Exception e, IXamlLineInfo lineInfo, Uri baseUri)
   at System.Windows.Markup.WpfXamlLoader.Load(XamlReader xamlReader, IXamlObjectWriterFactory writerFactory, Boolean skipJournaledProperties, Object rootObject, XamlObjectWriterSettings settings, Uri baseUri)
   at System.Windows.Markup.WpfXamlLoader.Load(XamlReader xamlReader, Boolean skipJournaledProperties, Uri baseUri)
   at System.Windows.Markup.XamlReader.Load(XamlReader xamlReader, ParserContext parserContext)
   at System.Windows.Markup.XamlReader.Load(XmlReader reader, ParserContext parserContext, XamlParseMode parseMode, Boolean useRestrictiveXamlReader)
   at System.Windows.Markup.XamlReader.Load(Stream stream, ParserContext parserContext, Boolean useRestrictiveXamlReader)
   at System.Windows.Markup.XamlReader.Load(Stream stream, ParserContext parserContext)
   at System.Windows.Xps.Packaging.XpsDocument.GetFixedDocumentSequence()
```

This appears to be a different issue to #575.

**Comment by weltkante** [comment]
This was already mentioned in the comments of #575, your scenario is probably caused by #585 but you'd have to check the inner exception to know for sure.

**Comment by elyoh** [comment]
Inner exception stack trace:

```
at MS.Internal.Text.TextInterface.Native.Util.ConvertHresultToException(Int32 hr)
   at MS.Internal.Text.TextInterface.Factory.CreateFontFace(Uri filePathUri, UInt32 faceIndex, FontSimulations fontSimulationFlags)
   at System.Windows.Media.GlyphTypeface.Initialize(Uri typefaceSource, StyleSimulations styleSimulations)
   at System.Windows.Media.GlyphTypeface..ctor(Uri typefaceSource, StyleSimulations styleSimulations)
   at System.Windows.Documents.Glyphs.ParseGlyphRunProperties()
   at System.Windows.Documents.Glyphs.ComputeMeasurementGlyphRunAndOrigin()
   at System.Windows.Documents.Glyphs.MeasureOverride(Size constraint)
   at System.Windows.FrameworkElement.MeasureCore(Size availableSize)
   at System.Windows.UIElement.Measure(Size availableSize)
   at System.Windows.Controls.Canvas.MeasureOverride(Size constraint)
   at System.Windows.FrameworkElement.MeasureCore(Size availableSize)
   at System.Windows.UIElement.Measure(Size availableSize)
   at System.Windows.Controls.Canvas.MeasureOverride(Size constraint)
   at System.Windows.FrameworkElement.MeasureCore(Size availableSize)
   at System.Windows.UIElement.Measure(Size availableSize)
   at System.Windows.Controls.Canvas.MeasureOverride(Size constraint)
   at System.Windows.FrameworkElement.MeasureCore(Size availableSize)
   at System.Windows.UIElement.Measure(Size availableSize)
   at System.Windows.Controls.Canvas.MeasureOverride(Size constraint)
   at System.Windows.FrameworkElement.MeasureCore(Size availableSize)
   at System.Windows.UIElement.Measure(Size availableSize)
   at System.Windows.Controls.Canvas.MeasureOverride(Size constraint)
   at System.Windows.FrameworkElement.MeasureCore(Size availableSize)
   at System.Windows.UIElement.Measure(Size availableSize)
   at System.Windows.Controls.Canvas.MeasureOverride(Size constraint)
   at System.Windows.FrameworkElement.MeasureCore(Size availableSize)
   at System.Windows.UIElement.Measure(Size availableSize)
   at System.Windows.Documents.FixedPage.MeasureOverride(Size constraint)
   at System.Windows.FrameworkElement.MeasureCore(Size availableSize)
   at System.Windows.UIElement.Measure(Size availableSize)
   at System.Windows.Documents.FixedDocument.GetPage(Int32 pageNumber)
   at System.Windows.Documents.FixedDocument.OnInitialized(Object sender, EventArgs e)
   at MS.Internal.Xaml.Runtime.ClrObjectRuntime.InitializationGuard(XamlType xamlType, Object obj, Boolean begin)
```

**Comment by weltkante** [comment]
That looks different from what I'd expect for #585 so probably needs a separate investigation.

**Comment by rladuca** [comment]
This seems like something specific to the XPS being displayed or something that has been fixed in Preview8.  

I tested this against several XPS documents sitting around in our test code and I was able to display them all without an error on `3.0.100-preview8-013275`.

Can you test against the latest daily build and, if this still reproduces, supply an XPS document that causes the issue to occur?

**Comment by rladuca** [comment]
https://github.com/dotnet/wpf/issues/585#issuecomment-503454479 seems to actually have a similar stack in Preview 8.  Using the document from that repro, I get:

```
   at System.Runtime.InteropServices.Marshal.ThrowExceptionForHR(Int32 errorCode, IntPtr errorInfo)
   at MS.Internal.Text.TextInterface.Native.Util.ConvertHresultToException(Int32 hr)
   at MS.Internal.Text.TextInterface.Factory.CreateFontFace(Uri filePathUri, UInt32 faceIndex, FontSimulations fontSimulationFlags)
   at System.Windows.Media.GlyphTypeface.Initialize(Uri typefaceSource, StyleSimulations styleSimulations)
   at System.Windows.Media.GlyphTypeface..ctor(Uri typefaceSource, StyleSimulations styleSimulations)
   at System.Windows.Documents.Glyphs.ParseGlyphRunProperties()
   at System.Windows.Documents.Glyphs.ComputeMeasurementGlyphRunAndOrigin()
   at System.Windows.Documents.Glyphs.MeasureOverride(Size constraint)
   at System.Windows.FrameworkElement.MeasureCore(Size availableSize)
   at System.Windows.UIElement.Measure(Size availableSize)
   at System.Windows.Documents.FixedPage.MeasureOverride(Size constraint)
   at System.Windows.FrameworkElement.MeasureCore(Size availableSize)
   at System.Windows.UIElement.Measure(Size availableSize)
   at System.Windows.Documents.FixedDocument.GetPage(Int32 pageNumber)
   at System.Windows.Documents.FixedDocument.OnInitialized(Object sender, EventArgs e)
   at System.Windows.FrameworkContentElement.RaiseInitialized(EventPrivateKey key, EventArgs e)
   at System.Windows.FrameworkContentElement.OnInitialized(EventArgs e)
   at System.Windows.FrameworkContentElement.TryFireInitialized()
   at System.Windows.FrameworkContentElement.EndInit()
   at MS.Internal.Xaml.Runtime.ClrObjectRuntime.InitializationGuard(XamlType xamlType, Object obj, Boolean begin)
```

@stevenbrix @vatsan-madhavan We should sync up on what we know from #585.

**Comment by rladuca** [comment]
This looks like our initialization of TypeFaces is directly calling into an unsupported part of the new zip implementation.

Take a look at this stack:

```
 	System.IO.Compression.dll!System.IO.Compression.DeflateStream.Position.get() Line 167	C#
 	System.IO.Packaging.dll!System.IO.Packaging.ZipWrappingStream.Position.get() Line 95	C#
>	PresentationCore.dll!MS.Internal.IO.Packaging.SynchronizingStream.Position.get()	Unknown
 	PresentationCore.dll!MS.Internal.IO.Packaging.DeobfuscatingStream.Read(byte[] buffer, int offset, int count)	Unknown
 	PresentationCore.dll!MS.Internal.FontCache.FontSource.StreamToByteArray(System.IO.Stream fontStream)	Unknown
 	PresentationCore.dll!MS.Internal.FontCache.FontSource.GetUnmanagedStream()	Unknown
 	DirectWriteForwarder.dll!MS.Internal.Text.TextInterface.FontFileStream.FontFileStream(MS.Internal.Text.TextInterface.IFontSource fontSource)	Unknown
 	DirectWriteForwarder.dll!MS.Internal.Text.TextInterface.FontFileLoader.CreateStreamFromKey(void* fontFileReferenceKey, uint fontFileReferenceKeySize, System.IntPtr* fontFileStream)	Unknown
 	[Native to Managed Transition]	
 	[Managed to Native Transition]	
 	DirectWriteForwarder.dll!MS.Internal.Text.TextInterface.Factory.GetFontCollection(System.Uri uri)	Unknown
 	PresentationCore.dll!MS.Internal.FontCache.DWriteFactory.GetFontCollectionFromFileOrFolder(System.Uri fontCollectionUri, bool isFolder)	Unknown
 	PresentationCore.dll!MS.Internal.FontCache.DWriteFactory.GetFontCollectionFromFile(System.Uri fontCollectionUri)	Unknown
 	PresentationCore.dll!System.Windows.Media.GlyphTypeface.Initialize(System.Uri typefaceSource, System.Windows.Media.StyleSimulations styleSimulations)	Unknown

```

I have a feeling this is going to affect any XPS with local font resources.  That's a pretty basic feature of XPS documents that won't work if we don't figure out some way around this.

**Comment by rladuca** [comment]
DeflateStream has [this code](https://github.com/dotnet/corefx/blob/cece0df13a74d188f96b08ae5af0cd4cd60f8f76/src/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs#L165):
```
        public override long Position
        {
            get { throw new NotSupportedException(SR.NotSupported); }
            set { throw new NotSupportedException(SR.NotSupported); }
        }
```

**Comment by rladuca** [comment]
Interestingly enough, we have a test XPS document that has embedded fonts, but the entire archive is a flat structure (the fonts are located in the root of the archive).  This file displays with no issue.

I have to examine what code paths are taken in that instance.

**Comment by stevenbrix** [comment]
The workaround here will be fairly straightforward and should follow what I did here: https://github.com/dotnet/wpf/pull/995/files#diff-7b801941d4352acdc7ea9eeaba96fecd

**Comment by rladuca** [comment]
@stevenbrix Oh I see, so, for instance, I just need to wrap the stream used [here](https://github.com/dotnet/wpf/blob/ac9d1b7a6b0ee7c44fd2875a1174b820b3940619/src/Microsoft.DotNet.Wpf/src/PresentationCore/MS/internal/FontCache/FontSource.cs#L205) instead of the direct cast?

**Comment by elyoh** [comment]
@rladuca The xps file does reference fonts on the local system.  It is a print out of an on screen view written in xaml.  I'll get a sample xps file uploaded on Friday.

**Comment by weltkante** [comment]
@rladuca these stack traces here from @elyoh and in #585 from @jbartlau look different. One stack trace is a managed NotSupportedException from `DeflateStream.get_Position()` but the other is a HRESULT native exception from `TextInterface.Factory.CreateFontFace`. I'm not sure they have the same root cause, the stack traces sure don't look that way.

**Comment by rladuca** [comment]
@weltkante The second stack I posted is the actual exception stack from loading @jbartlau's file with first chance exceptions turned on.  Loading the font when the font is located within the XPS always tries to access position, assuming it doesn't live in the root of the archive.  Take a look at the offending code in [DeobfuscatingStream](https://github.com/dotnet/wpf/blob/08128b97bedfa40030d30db64dce7d5ea0924fc9/src/Microsoft.DotNet.Wpf/src/PresentationCore/MS/internal/IO/Packaging/DeobfuscatingStream.cs#L56) and you'll see what I am talking about.

I'm pretty sure the root cause of both issues is exactly the same even if the various stacks are slightly different in how we get there.

**Comment by rladuca** [comment]
@weltkante Ah, I see where I might have conflated the two.  I assumed the font loading path would get hit both ways.  I'll wait for @elyoh's reproduction XPS first.

**Comment by elyoh** [comment]
Here is a sample xps file which leads to the error.

[2069bGRVwEKTGite3tlNJw.zip](https://github.com/dotnet/wpf/files/3411308/2069bGRVwEKTGite3tlNJw.zip)

I tried using the same file in 3.0.100-preview8-013330 and received the same error.

**Comment by rladuca** [comment]
@elyoh Thanks!  I'll have a look today and see what I can figure out.

**Comment by rladuca** [comment]
@elyoh @weltkante This actually turned out to be the same error, just caught in a slightly different manner.

When running with my recent fixes for #585, the document displays correctly.

![image](https://user-images.githubusercontent.com/20006987/61565226-42a6f900-aa2d-11e9-9aef-eaf56a558566.png)

When running without those changes, I get the same error on accessing DeflateStream.Position.  The difference here is the call stack that leads to the error has an error trap that re-throws.

See [here](https://github.com/dotnet/wpf/blob/e4a212b4fd7d3bb6bce7d88cf6f0471a9fc8e6f4/src/Microsoft.DotNet.Wpf/src/DirectWriteForwarder/CPP/DWriteWrapper/FontFileLoader.cpp#L50).

In any case, the same workaround will help with your issue as I gave in #585 , open the XPS as read/write.

Fixed by #1311
